docs/core/v3.0.rst

```diff
@@ -0,0 +1,1762 @@
+
+The Zarr core development team maintains a repository of codec
+specifications, which are hosted alongside this specification in the
+`zarr-specs GitHub repository`_, and which are
+published on the `zarr-specs documentation Web site
+<http://zarr-specs.readthedocs.io/>`_. For ease of discovery, it is
+recommended that codec specifications are contributed to the
+zarr-specs GitHub repository. However, codec specifications may be
+maintained by any group or organisation and published in any location
+on the Web. For further details of the process for contributing a
+codec specification to the zarr-specs GitHub repository, see the Zarr
+community process specification.
+
+Further details of how a compressor is configured for an array are
+given in the section below on `Array metadata`_.
+
+
+Metadata
+========
+
+This section defines the structure of metadata documents for Zarr hierarchies,
+which consists of three types of metadata documents: an entry point metadata
+document (``zarr.json``), array metadata documents, and group metadata
+documents. Each type of metadata document is described in the following
+subsections.
+
+Metadata documents are defined here using the JSON
+type system defined in [RFC8259]_. In this section, the terms "value",
+"number", "string" and "object" are used to denote the types as
+defined in [RFC8259]_. The term "array" is also used as defined in
+[RFC8259]_, except where qualified as "Zarr array". Following
+[RFC8259]_, this section also describes an object as a set of
+name/value pairs. This section also defines how metadata documents are
+encoded for storage.
+
+
+Only the top level metadata document ``zarr.json`` is guaranteed to be of JSON
+type, and can be used to define other formats for array-level and group-level
+metadata documents. In the case where non-JSON metadata documents are used in a
+Zarr hierarchy, the following sections on group and array level metadata are
+non-normative, but other metadata formats are expected to define some
+equivalence relations with the JSON documents.
+
+
+Entry point metadata
+--------------------
+
+Each Zarr hierarchy must have an entry point metadata document, which
+provides essential information regarding the format version being
+used, the encoding being used for group and array metadata, and any
+extensions that affect the layout or interpretation of data
+in the store.
+
+The entry point metadata document must contain a single object
+containing the following names:
+
+``zarr_format``
+^^^^^^^^^^^^^^^
+
+    A string containing the URI of the Zarr core
+    specification that defines the metadata format. For Zarr
+    hierarchies conforming to this specification, the value must be
+    the string "https://purl.org/zarr/spec/core/3.0".
+
+    Implementations of this specification may assume that the final path
+    segment of this URI ("3.0") represents the core specification version
+    number, where "3" is the major version number and "0" is the minor
```

Comment:
It is not clear how the "minor" version number will be used.  How do we anticipate an implementation making use of the minor version number?  Perhaps instead we should eliminate it and just rely on "extensions" for added features?

Additionally, there is the issue that this version number is written in the "entry point metadata document".  But someone might, for example, first create a zarr root with one program, and then later write a new array with a different program that supports an earlier or later version of zarr 3.  How would the root version be updated in that case?

Reply:
How about we adopt semantic versioning as per https://semver.org/ ?

> MAJOR version when you make incompatible API changes
MINOR version when you add functionality in a backwards compatible manner
PATCH version when you make backwards compatible bug fixes

Reply:
Since we are talking about a file format specification rather than a piece of software I think it is important to clarify exactly what would constitute a change in MAJOR, MINOR, and PATCH and how implementations would make use of that information.

In particular, the key cases are:
1. Implementation (i.e. a specific version of the implementation) encounters a zarr repository that claims to use a version of the spec that was already published before the implementation was written.  In that case the implementation can have any number of special cases as required and it doesn't really matter what the version number scheme is.
2. Implementation encounters a zarr repository that claims to use a version of the spec that was *not* published when the implementation was written.  In this case the implementation must decide to either:
  (a) Proceed and treat it the same as the latest version that is supported, possibly ignoring any additional metadata keys that are present but not defined in the latest supported version of the spec.
  (b) Fail with an error.

Presumably an increased MAJOR version leads to (b) while a change to MINOR or PATCH would lead to (a).   But then it is not clear what the purpose is of even including the MINOR or PATCH number in the file format itself since the implementation will ignore it.

I would propose instead that the metadata itself store only the MAJOR version, i.e. "3".  Instead of "3" it could also be some URL that points to the latest version 3 specification.

When publishing the spec itself, it could additionally include a MINOR and PATCH version number.  These would be useful in discussing the spec but would not be part of the metadata.  We might increase MINOR if any ignorable changes to the stored representation have been made (basically equivalent to a `must_understand=False` extension), and increase PATCH when only editorial changes have been made.  But I'm also not sure how useful it would be to distinguish between MINOR and PATCH.

Reply:
Also linking to the [Stability Policy](section), which was also discussed in those comments: https://github.com/zarr-developers/zarr-specs/pull/142#discussion_r870513234

Reply:
> It is not clear how the "minor" version number will be used. How do we anticipate an implementation making use of the minor version number? Perhaps instead we should eliminate it and just rely on "extensions" for added features?

I agree that there's currently no clear use for the minor version except to distinguish between different versions of the same compatible document. This seems useful to version the spec itself, but might not be important for the metadata entry in zarr_format. But also, I don't see a problem to also have it there, especially to have a clear identifier for the spec.

> Additionally, there is the issue that this version number is written in the "entry point metadata document". But someone might, for example, first create a zarr root with one program, and then later write a new array with a different program that supports an earlier or later version of zarr 3. How would the root version be updated in that case?

As mentioned above, that shouldn't be important IMO, and I guess an implementation may choose to update the version or not :shrug:

I'm marking this as resolved in this PR to better keep an overview, as there's no immediate actionable. If you'd like to see changes to the spec in regard of the zarr core version, please open an issue about it or even better a PR.